cvmfs/sqlitemem.h

```diff
@@ -0,0 +1,291 @@
+  /**
+   * In total: 64kB scratch memory
+   */
+  static const unsigned kScratchSize = kScratchSlotSize * kScratchNoSlots;
+
+  /**
+   * Empricially, the largest page cache allocation is 1296B.
+   */
+  static const unsigned kPageCacheSlotSize = 1300;
+  /**
+   * Number of pages that can be cached.
+   */
+  static const unsigned kPageCacheNoSlots = 4000;
+  /**
+   * In total: ~5MB
+   */
+  static const unsigned kPageCacheSize = kPageCacheSlotSize * kPageCacheNoSlots;
+
+  /**
+   * 32 bytes per slot is an empirically good value so that memory is not wasted
+   * (too much) and many allocations fit within the boundary.
+   */
+  static const unsigned kLookasideSlotSize = 32;
+  /**
+   * 128 slots with 32bytes accumulate to 4kB.  See LookasideBufferArena below.
+   */
+  static const unsigned kLookasideSlotsPerDb = 128;
+
+
+  /**
+   * An mmap'd block of general purpose memory for malloc/free in sqlite.  Uses
+   * the "boundary-tag system" as described in TAOCP.
+   * Not thread-safe.  Pointers are not aligned.
+   */
+  class MallocArena {
+   public:
+    /**
+     * Should be larger than 10 times the largest allocation, which for reading
+     * sqlite file catalogs is 64kB.  An arena size of 8MB limits the total
+     * number of arenas (mapped blocks) to <40, given typical storage needs for
+     * 10,000 open catalogs.
+     * Has to be a power of 2MB.
+     */
+    static const unsigned kArenaSize = 8 * 1024 * 1024;
+
+    /**
+     * Avoid very small free blocks.  This must be anyway larger than
+     * sizeof(AvailBlockCtl) + sizeof(AvailBlockTag).
+     */
+    static const int kMinBlockSize = 20;
+
+    /**
+     * Returns the MallocArena that houses the destination of ptr.
+     */
+    static inline MallocArena *GetMallocArena(void *ptr) {
+      void *arena = reinterpret_cast<void *>(
+        uintptr_t(ptr) & ~(uintptr_t(kArenaSize) - uintptr_t(1)));
+      return *reinterpret_cast<MallocArena **>(arena);
+    }
+
+    MallocArena();
+    ~MallocArena();
+
+    void *Malloc(const uint32_t size);
+    void Free(void *ptr);
+    inline bool Contains(void *ptr) const {
+      return GetMallocArena(ptr) == this;
+    }
+    uint32_t GetSize(void *ptr) const;
+    bool IsEmpty() const { return no_reserved_ == 0; }
+
+   private:
+    static const char kTagAvail = 0;
+    static const char kTagReserved = 1;
+
+    /**
+     * Lower boundary of a free block.  Note that the linking of blocks is not
+     * necessarily in ascending order but random.
+     */
+    struct AvailBlockCtl {
+      AvailBlockCtl *GetNextPtr(char *base) {
+        return reinterpret_cast<AvailBlockCtl *>(base + link_next);
+      }
+      AvailBlockCtl *GetPrevPtr(char *base) {
+        return reinterpret_cast<AvailBlockCtl *>(base + link_prev);
+      }
+      int32_t ConvertToLink(char *base) {
+        return reinterpret_cast<char *>(this) - base;
+      }
+      int32_t size;  // always positive
+      int32_t link_next;
+      int32_t link_prev;
```

Comment:
They are offsets.  Because the arena is small, 4 bytes suffice instead of 8 bytes for pointers.